test/test_selectors.py

```diff
@@ -0,0 +1,670 @@
+
+        key = s.register(rd, selectors.EVENT_READ, "data")
+        self.assertEqual(key, s.get_key(rd))
+
+        # Unknown fileobj
+        self.assertRaises(KeyError, s.get_key, 999999)
+
+    def test_get_map(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        keys = s.get_map()
+        self.assertFalse(keys)
+        self.assertEqual(len(keys), 0)
+        self.assertEqual(list(keys), [])
+        key = s.register(rd, selectors.EVENT_READ, "data")
+        self.assertIn(rd, keys)
+        self.assertEqual(key, keys[rd])
+        self.assertEqual(len(keys), 1)
+        self.assertEqual(list(keys), [rd.fileno()])
+        self.assertEqual(list(keys.values()), [key])
+
+        # Unknown fileobj
+        self.assertRaises(KeyError, keys.__getitem__, 999999)
+
+        # Read-only mapping
+        with self.assertRaises(TypeError):
+            del keys[rd]
+
+    def test_register(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        # Ensure that the file is not yet added.
+        self.assertEqual(0, len(s.get_map()))
+        self.assertRaises(KeyError, lambda: s.get_map()[rd.fileno()])
+        self.assertRaises(KeyError, s.get_key, rd)
+        self.assertEqual(None, s._key_from_fd(rd.fileno()))
+
+        data = object()
+        key = s.register(rd, selectors.EVENT_READ, data)
+        self.assertIsInstance(key, selectors.SelectorKey)
+        self.assertEqual(key.fileobj, rd)
+        self.assertEqual(key.fd, rd.fileno())
+        self.assertEqual(key.events, selectors.EVENT_READ)
+        self.assertIs(key.data, data)
+        self.assertEqual(1, len(s.get_map()))
+        for fd in s.get_map():
+            self.assertEqual(fd, rd.fileno())
+
+    def test_register_bad_event(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        self.assertRaises(ValueError, s.register, rd, 99999)
+
+    def test_register_negative_fd(self):
+        s = self.make_selector()
+        self.assertRaises(ValueError, s.register, -1, selectors.EVENT_READ)
+
+    def test_register_invalid_fileobj(self):
+        s = self.make_selector()
+        self.assertRaises(ValueError, s.register, "string", selectors.EVENT_READ)
+
+    def test_reregister_fd_same_fileobj(self):
+        s, rd, wr = self.standard_setup()
+        self.assertRaises(KeyError, s.register, rd, selectors.EVENT_READ)
+
+    def test_reregister_fd_different_fileobj(self):
+        s, rd, wr = self.standard_setup()
+        self.assertRaises(KeyError, s.register, rd.fileno(), selectors.EVENT_READ)
+
+    def test_context_manager(self):
+        s = self.make_selector()
+        rd, wr = self.make_socketpair()
+
+        with s as sel:
+            sel.register(rd, selectors.EVENT_READ)
+            sel.register(wr, selectors.EVENT_WRITE)
+
+        self.assertRaises(RuntimeError, s.get_key, rd)
+        self.assertRaises(RuntimeError, s.get_key, wr)
```

Comment:
We may want to assert that `get_key` *does* return what we expect inside the context manager.